### PR TITLE
Do partition assignment using pod name and total block builders and partitions

### PR DESCRIPTION
**Pull request by codesome**
This move is in order to make partitions stable in a block builder to most extent.
This is not aimed to be the final design, but a temporary stop-gap to help test block builders (although this is a simplified version of a proposed design).
Note: If we plan to keep it this way, we need to make identifying the number of block builders and partitions automatic, using some discovery mechanism.

This PR also includes https://github.com/grafana/mimir/pull/8804

### CODE REVIEW

File: pkg/blockbuilder/blockbuilder.go [hunk 1]
```diff
@@ -92,134 +123,211 @@ func (b *BlockBuilder) starting(ctx context.Context) (err error) {
...
+		opts = append(
+			commonKafkaClientOptions(b.cfg.Kafka, b.logger, nil),
+			opts...,
+		)
+		kafkaClient, err := kgo.NewClient(opts...)
+		if err != nil {
+			return fmt.Errorf("creating kafka client: %w", err)
+		}
 
+		b.kafkaClients[int32(part)] = kafkaClient
```
[the remark this answers is not in the document]
Comment (codesome) [reply pkg/blockbuilder/blockbuilder.go:194]: For this prototype we can probably get away with a single client. But later, can we change the partitions we consume from within the client?

Also, can we pause and resume fetching from partitions like before?

[the remark this answers is not in the document]
Comment (narqo) [reply pkg/blockbuilder/blockbuilder.go:194]: > Also, can we pause and resume fetching from partitions like before?

I don't think pausing stops the client from polling the metadata from the broker. Or does it?

[the remark this answers is not in the document]
Comment (codesome) [reply pkg/blockbuilder/blockbuilder.go:194]: > I don't think pausing stops the client from polling the metadata from the broker (that was the problem in https://github.com/grafana/mimir/pull/8088). Or does it?

My question was more about when we have a single client, to know if we can still control which partitions we consume when we have a single client. I guess it works similar to how we have before this PR, so I am guessing we can still control pausing and resuming individual partitions. I will implement a single client.

**Review by narqo — APPROVED** [review]
🔥 works for me. We'll revisit the rest of the details and do clean up, as a follow-up work.

**Pull request merged**